### ISSUE TITLE
privacy-preserving multisig transactions

### ISSUE THREAD

**Issue by daira** [zcash/zcash#782]
I don't know how to do privacy-preserving multisig at the moment, so this ticket is quite speculative. This feature would allow you to combine several independently generated proofs of ownership of distinct spending keys in order to be able to spend coins sent to a private multisig address, without requiring those keys to be brought together on the same machine.

The following desirable properties of a solution seem to be in conflict:
* no significant increase in the cost of non-multisig Pours;
* indistinguishability between Pours involved in multisig transactions vs non-multisig transactions.

**Comment by daira** [comment]
Of course, you can publish a transparent multisig address and then, by convention:
* only spend to that address using transactions which take the whole value from the public output of a Pour;
* only spend from that address using transactions which transfer the whole value to the public input of a Pour.

Let's call this the "multisig Pour sandwich" approach. While it does successfully hide the source and destination addresses, it fails to hide the fact that a transparent value has been sent to the multisig address at a particular time. Also, it relies on transport-level anonymity (Tor) to hide any correlation between that fact and the transport layer source of the sending and receiving transactions.

**Comment by nathan-at-least** [comment]
@daira, I think this "sandwich approach" will be a general pattern for many script applications not (yet) directly supported by the snark circuit.

A low-overhead circuit design sounds like a fun problem. I'm excited to launch so we can then switch to some 2.0 speculative design hackfests. ;-)

**Comment by daira** [comment]
It might be more feasible to make an N-key Pour look indistinguishable to N chained Pours, rather than to a single Pour.

**Comment by daira** [comment]
I had a half-formed idea about how to do this by making an address commit to a public key for a threshold blind/key-private signature scheme. A signature in this scheme would be key-private, but could be input to the SNARK so that it can check somehow that it is for the right address. The difficult part would be finding a scheme that allows the public key operations to be done outside the SNARK, while still being key-private.

**Comment by daira** [comment]
Oops, pressed the wrong button :-)

**Comment by daira** [comment]
[OpenBazaar](https://blog.openbazaar.org/what-is-openbazaar/) is a good motivating example for privacy-preserving multisig: it uses that to allow a moderator for each transaction to resolve disputes between buyer and seller.

**Comment by daira** [comment]
Since the JoinSplit circuit is fixed-size, it is clear that if we want to support an arbitrary number of signatures in a multisig transaction without increasing the circuit size for the common case, they must be represented using multiple JoinSplits.

Here's one way this could work, with only a little handwaving. Suppose that each JoinSplit contributes one "signature". For z-addresses, rather than a conventional signature, the zk-SNARK proves knowledge of the spending key a<sub>sk</sub> for a given paying key a<sub>pk</sub>. Denote an _n_-of-_m_ multisig address as MultiAddr(n, a<sub>pk,1.._m_</sub>). Modify the note commitment scheme so that it is possible to commit either to an ordinary note as before, or to a "partial note" which records in a bit field which "signatures" have already been made. Paying to a multisig address outputs a partial note with no signatures. This partial note can be used as input to another JoinSplit that proves knowledge of a spending key a<sub>pk,_i_</sub>, and commits to another partial note with that "signature" bit _i_ filled in. After _n_ JoinSplits have proven knowledge of _n_ distinct spending keys, the last such JoinSplit can spend the note.

A partial note has to record *which* signatures have already been made, in order to prevent just using the same spending key _n_ times. So the commitment input must include a bit field of size at least _m_. Strictly speaking this puts a limit on _m_, but the circuit cost is only at most _m_ / (input block size of hash in bits) hash compressions. For SHA256Compress the input block size is 512 bits, so we can support _m_ up to 512 with only one extra block (per input or output that is allowed to be used in a multisig transaction). Alternatively if we don't need to support _m_ as large as this, we may be able to fit the bit field in an existing input block of the note commitment hash.

**Comment by daira** [comment]
Bitcoin supports up to _m_ = 20 in theory, but only _m_ = 3 in standard transactions if you do it in the obvious way: https://bitcoin.stackexchange.com/questions/23893/what-are-the-limits-of-m-and-n-in-m-of-n-multisig-addresses

I'm somewhat surprised by the latter. Apparently there is a hack, linked on that StackExchange page, to support up to 8-of-15 in standard transactions.

**Comment by daira** [comment]
I put scare quotes around the term 'signature' in https://github.com/zcash/zcash/issues/782#issuecomment-285244846 , but that's not necessary: a NIZK argument of knowledge of the preimage of a hash is a signature, by the usual security/correctness definitions for signature schemes.

**Comment by daira** [comment]
We can simplify this scheme conceptually: instead of thinking of creating a partial note that records which signatures have been made, we can think of creating an ordinary note that is sent to another multisig address that records which signatures have been made. I.e. suppose the input note comes from MultiAddr(n, a<sub>pk,M</sub>, pk<sub>enc</sub>), where M is a subset of {1.._m_} and the JoinSplit proves knowledge of a<sub>sk,_i_</sub> for some _i_ ∈ M. Then we think of the note as being sent to MultiAddr(n-1, a<sub>pk,M \ {_i_}</sub>, pk<sub>enc</sub>).

This doesn't actually change how the circuit would need to work: the note commitments have to include the bitfield of used signatures and the output multisig address in either case, so it's just a matter of how you represent that in the note commitment input.

Speaking of which, I glossed over an important detail in how multisig addresses are represented in note commitments. In the current Zcash protocol, a note commitment commits to (a<sub>pk</sub>, v, ρ). For multisig, the a<sub>pk</sub> has to be replaced with some representation of M and a<sub>pk,M</sub>, i.e. the paying keys of a multisig address that have not yet been authorized by a signature. This consists of up to _m_ hashes. We also have to be able to prove in the circuit that a<sub>sk,_i_</sub> is a PRF<sup>addr</sup>-preimage of a<sub>pk,_i_</sub>. The obvious way to do so is to represent a<sub>pk,1.._m_</sub> as the leaves of a Merkle tree; then the circuit can index into this tree to find a<sub>pk,_i_</sub>. This also allows the circuit to check that the output multisig address is correctly related to the input multisig address: it has the same Merkle root hash, but index _i_ in the bitfield representing M indicates that knowledge of a<sub>sk,_i_</sub> has already been proven.

The drawback of this is that we need O(log(_m_<sub>max</sub>)) circuit hashes to index into the Merkle tree. In practice _m_<sub>max</sub> = 8 would probably be sufficient, but 3 extra hashes for a feature that won't always be used is not ideal.

**Comment by daira** [comment]
Here's a more precise performance analysis. Assume we are doing #647, and assume d<sub>Merkle</sub> = 29 and _m_<sub>max</sub> = 8. Then the baseline number of circuit hashes is 40 (29×MerkleCRH, 1×PRF<sup>nf</sup>, 2×PRF<sup>ρ</sup>, 1×PRF<sup>addr</sup>, 1×PRF<sup>pk</sup>, 2 for the input NoteCommitment, 4 for the output NoteCommitments).

The proposal above requires log<sub>2</sub>(_m_<sub>max</sub>) = 3 extra MerkleCRH hashes (per input, but there is only one input), so the additional circuit cost is approximately 7.5%. It would probably be a little more than this due to bookkeeping such as checking that the output multisig address is correctly related to the input one when not all signatures have been provided, and the addition of _i_ and another Merkle path to the witness.

 (There is no increase in the number of hashes needed for NoteCommitments because we only need 3 extra bits to represent M, and we have 15 bytes spare in the last block of input to NoteCommitment; the Merkle tree root of the multisig address replaces a<sub>pk</sub>.)

**Comment by daira** [comment]
If we also do https://github.com/zcash/zcash/issues/2171#issuecomment-285963483 , then the extra 3 MerkleCRH hashes go into the input section, so we have 11 hashes there instead of 8. However there's a problem with making sure that the output commitment matches the input one: the only way I can think of to do this requires the input circuit to also compute what the note commitment of the partially spent note should be (and then convey that with the value commitment), which is an extra 2 hashes that weren't required with the monolithic circuit.

**Comment by daira** [comment]
Ah there's an elegant trick that allows us to do without the bitfield M.

The input side of a JoinSplit publishes a nullifier nf that is computed as PRF<sup>nf</sup><sub>a<sub>sk,_i_</sub></sub>(ρ). We can use this to cause repeated use of the same private key for a multisig note to be treated as an attempted double-spend, thus avoiding the need to explicitly keep track of which private keys have been used. This requires ρ to be the same for the partially spent output note as for the input note, but that is not a problem in this case.

We add _n_, the number of signatures that are still required, to the note commitment in place of M. Note that the size of _n_ is log<sub>2</sub>(_m_<sub>max</sub>) bits, and replaces M which was _m_<sub>max</sub> bits &mdash; so we can support a larger _m_<sub>max</sub> without having to increase the size of the note commitment input too much. Now, we enforce that for a partial spend, _n_ is decreased by 1, and (a<sup>root</sup><sub>pk</sub>, v, ρ) are the same (instead of requiring ρ to be computed by PRF<sup>ρ</sup>). Here a<sup>root</sup><sub>pk</sub> is the root of a Merkle tree over a<sub>pk,1.._m_</sub>. When _n_ gets to 1 on the input side, the note can be spent normally.

Another thing I hadn't noticed is that a multisig z-address only requires (_n_, a<sup>root</sup><sub>pk</sub>, pk<sub>enc</sub>); it does not require a<sub>pk,1.._m_</sub>. So it doesn't have to be much larger than an ordinary z-address.

**Comment by daira** [comment]
Oops, that would allow you to double-spend a note via disjoint subsets of spending keys, if _m_ >= 2_n_.

**Comment by daira** [comment]
A variation of the above problem also occurs for the bitfield approach. To solve it, there can only be a single nullifier for a given commitment. This means that the a<sub>sk</sub> used to compute the nullifier for a given signing cannot be the subkey used for that signing; it must be a separate key. The following works:

![multisig](https://cloud.githubusercontent.com/assets/643204/23931092/2dbba44e-0928-11e7-9503-5c4c377c9060.png)

In this diagram, lines represent hash inputs going upwards, e.g. a<sub>pk</sub> is a hash of a<sub>sk</sub> and a<sub>rk</sub>, and the lilac triangle is a Merkle tree. To make a partial spend, you have to prove knowledge of both a<sub>sk</sub> and one of the a<sub>msk,i</sub>. This allows fixing the double-spend vulnerability by ensuring that each signature constructs an output note commitment with a new ρ, and then there is only one nullifier that will work when performing the next signature.

**Comment by daira** [comment]
Here is the proposed JoinSplit statement, including #647. For N<sup>new</sup> = 2 it has 13 + d<sub>Merkle</sub> + log<sub>2</sub>(_m_) hashes, which is log<sub>2</sub>(_m_) more than the #647 circuit without multisig.

"≛" means "is constrained equal to".

statement JoinSplit(
    rt ⦂ 𝔹<sup>[ℓ<sub>Merkle</sub>]</sup>,
    nf<sup>old</sup> ⦂ 𝔹<sup>[ℓ<sub>PRF</sub>]</sup>,
    cm<sup>new</sup><sub>1..N<sup>new</sup></sub> ⦂ COMM.Output<sup>[N<sup>new</sup>]</sup>,
    h<sub>Sig</sub> ⦂ 𝔹<sup>[ℓ<sub>hSig</sub>]</sup>,
    h ⦂ 𝔹<sup>[ℓ<sub>PRF</sub>]</sup>,
    cv<sup>pub</sup><sub>old</sub> ⦂ COMM.Output,
    cv<sup>pub</sup><sub>new</sub> ⦂ COMM.Output,
    private path ⦂ 𝔹<sup>[ℓ<sub>Merkle</sub>][d<sub>Merkle</sub>]</sup>,
    private **n**<sup>old</sup> ⦂ Note,
    private a<sup>old</sup><sub>sk</sub> ⦂ 𝔹<sup>[ℓ<sub>a<sub>sk</sub></sub>]</sup>,
    private a<sup>old</sup><sub>msk</sub> ⦂ 𝔹<sup>[ℓ<sub>a<sub>sk</sub></sub>]</sup>,
    private multiSigPath ⦂ 𝔹<sup>[ℓ<sub>Merkle</sub>][d<sub>_m_</sub>]</sup>,
    private **n**<sup>new</sup><sub>1..N<sup>new</sup></sub> ⦂ Note[N<sup>new</sup>],
    private φ ⦂ 𝔹<sup>[ℓ<sub>φ</sub>]</sup>,
    private enforceMerklePath ⦂ 𝔹,
    private j ⦂ {0..N<sup>new</sup>}
    private v<sup>pub</sup><sub>old</sub> ⦂ {0..2<sup>64</sup>-1},
    private v<sup>pub</sup><sub>new</sub> ⦂ {0..2<sup>64</sup>-1},
    private r<sup>pub</sup><sub>old</sub> ⦂ COMM.Trapdoor,
    private r<sup>pub</sup><sub>new</sub> ⦂ COMM.Trapdoor,
) {
    let (a<sup>old</sup><sub>pk</sub>, v<sup>old</sup>, ρ<sup>old</sup>, r<sup>old</sup>, _n_<sup>old</sup>) = **n**<sup>old</sup>

    // Input commitment integrity.
    cm<sup>old</sup> ≛ NoteCommitment(**n**<sup>old</sup>)

    for i ∈ {1..N<sup>new</sup>} {
        let (a<sup>new</sup><sub>pk,i</sub>, v<sup>new</sup><sub>i</sub>, ρ<sup>new</sup><sub>i</sub>, r<sup>new</sup><sub>i</sub>, _n_<sup>new</sup><sub>i</sub>) = **n**<sup>new</sup><sub>i</sub>

        // Output commitment integrity.
        cm<sup>new</sup><sub>i</sub> ≛ NoteCommitment(**n**<sup>new</sup><sub>i</sub>)
    }

    // Merkle path validity. (For simplicity we don't include #2171 or #43 here.)
    rt ≛ MerkleRoot(path, cm<sup>old</sup>)

    // Merkle path enforcement.
    if (v<sup>old</sup> ≠ 0) {
        enforceMerklePath ≛ 1
    }

    // Balance.
    if (_n_<sup>old</sup> > 1) {
        // Non-final multisig. The output note (indexed j) must use the same (a<sub>pk</sub>, v, ρ), and requires
        // one fewer signature. The input note provides no value, since it has not been fully spent.
        j ≛̸ 0
        **n**<sup>new</sup><sub>j</sub> ≛ (a<sup>old</sup><sub>pk</sub>, v<sup>old</sup>, ρ<sup>old</sup>, r<sup>new</sup><sub>j</sub>, _n_<sup>old</sup> - 1)
        let v<sup>in</sup> = 0
    } else {
        // Final multisig or non-multisig. The input note provides value.
        j ≛ 0
        let v<sup>in</sup> = v<sup>old</sup>
    }
    v<sup>pub</sup><sub>old</sub> + v<sup>in</sup> ≛ v<sup>pub</sup><sub>new</sub> + ∑[v<sup>new</sup><sub>i</sub> for i from 1 up to N<sup>new</sup>]

    // Nullifier integrity.
    nf<sup>old</sup> ≛ PRF<sup>nf</sup><sub>a<sup>old</sup><sub>sk</sub></sub>(ρ<sup>old</sup>)

    // Spend authority.
    if (_n_<sup>old</sup> > 0) {
        // Multisig spend. multiSigPath identifies which subkey is used.
        a<sup>old</sup><sub>pk</sub> ≛ PRF<sup>maddr</sup><sub>a<sub>sk</sub></sub>(MerkleRoot(multiSigPath, a<sup>old</sup><sub>msk</sub>))
    } else {
        a<sup>old</sup><sub>pk</sub> ≛ PRF<sup>addr</sup><sub>a<sub>sk</sub></sub>(0)
    }

    // Nonmalleability.
    h ≛ PRF<sup>pk</sup><sub>a<sup>old</sup><sub>sk</sub></sub>(hSig)

    // Uniqueness of ρ<sup>new</sup><sub>1..N<sup>new</sup> \\ j</sub>.
    for i ∈ {1..N<sup>new</sup>} \\ {j} {
        ρ<sup>new</sup><sub>i</sub> ≛ PRF<sup>ρ</sup><sub>φ</sub>(i, hSig)
    }

    // Commitment to incoming value.
    cv<sup>pub</sup></sub>old</sub> ≛ COMM<sub>r<sup>pub</sup><sub>old</sub></sub>(v<sup>pub</sup><sub>old</sub>)

    // Commitment to outgoing value.
    cv<sup>pub</sup></sub>new</sub> ≛ COMM<sub>r<sup>pub</sup><sub>new</sub></sub>(v<sup>pub</sup><sub>new</sub>)
}

**Comment by daira** [comment]
The multisig address derivation could use a ternary Merkle tree as described in #2173. This would allow up to _m_<sub>max</sub> = 27 with three SHA256Compress hashes, rather than _m_<sub>max</sub> = 8 for the same number of hashes.

**Comment by arielgabizon** [comment]
Here's an idea I've been playing around with that I like.
Like Nathan suggested it works for general scripts.
I should go over the details of @daira's scheme,
but anyway the advantage here is that it would require only a consensus change, and not a circuit change.
The disadvatnage is that in case of multisig that signatures and signers' public keys would be revealed , but sender, recipient, amount would be secret.

We add a new type of tx, call it JSp2sh  (JS for Jointsplit).
The tx contains a  p2sh scriptPubKey, and a JoinSplit.
It's identified by an address prefix that makes the nodes do the following.
- Add the nullfiers to the nullifier set
- Do *not* add the new note commitments to the NC tree.

Then to "redeem" the Join-Split, one must supply the matching scriptSig.
The consensus rules  will say that when a valid transaction from this address
is sent the NC's in the Join-Split are to be added to the NC tree.

This way, you could get a Join-Split whose spending requires an arbitrary bitcoin script.
Again, with the disadvantage that the scriptSig is public but with no required circuit change.

**Comment by arielgabizon** [comment]
@daira  I'm trying to understand the circuit.. I don't really get it. The n out of m signing parties need to prove they know a path to a_rk? what if their n leaves don't consitute a subtree - i.e. don't contain a path to the root?
Also, the spend authority check should only be done when j=0? i.e. when it's a final multisig with all the signatures?

**Comment by daira** [comment]
Yes, each signing party must demonstrate that it knows one of the a<sup>old</sup><sub>msk,i</sub> &ndash;that has not already been used&ndash; and that it knows a path from it to a<sub>rk</sub>. The signing party must also prove it knows a<sub>sk</sub>.

> what if their n leaves don't consitute a subtree - i.e. don't contain a path to the root?

The path from a given a<sup>old</sup><sub>msk,i</sub> is part of that private key. Each of the n parties independently proves they know one of the a<sup>old</sup><sub>msk,i</sub>. Does that answer your question?

> Also, the spend authority check should only be done when j=0?

The spend authority check must always be done, because there must be exactly one nullifier, PRF<sup>nf</sup><sub>a<sup>old</sup><sub>sk</sub></sub>(ρ<sup>old</sup>), that is revealed when a given partial note is spent. (If not, then you could double-spend a partial note.) If the nullifier were computed based on a<sup>old</sup><sub>msk,i</sub> then there would be more than one possibility &mdash; therefore it must be based on a<sub>sk</sub>, which is why the signing party must prove that they know that.

Note that we have another candidate for how to support multisig using threshold signatures, as mentioned in #2227. That alternative has not been fully fleshed out yet; I don't know whether it can support multisig spends and conventional spends indistinguishably.

**Comment by arielgabizon** [comment]
OK that makes more sense to me now. I'm wondering about the following: Do you need a different subcircuit for each j in [m], to check if there's a path from the j'th leaf to the root in this tree with root a<sub>rk<\sub>

rather than just checking if there is a path from some leaf to the root

**Comment by daira** [comment]
No, we want there to be only one circuit. This is no different from the check that a note commitment exists in the Merkle tree (we can use the same gadget).

**Comment by miguelmarco** [comment]
Would it be possible to "fake" the 2-of-2 multisig behaviour in the following way?:

Each party (let's call them A and B) of the 2 chooses a secret, and then combine those secrets to create a common public key (via a Diffie-Helman interchange or something like that). So they have a common public key, and can use it to encode a z-address. This z-address is going to play the role of a multisig address.

When they agree that the fundings in the multisig should go to (say) B, A sends the secret, so B can reconstruct the private key and spend it. 

Does this make sense?

**Comment by daira** [comment]
https://freedom-to-tinker.com/2017/09/11/blocksci-a-platform-for-blockchain-science-and-exploration/

"Exploratory graph analysis using BlockSci allowed us to discover a behavioral pattern in the usage of multisignatures that weakens security. Multisignatures are a security-enhancing mechanism that distribute control of an address over a number of different public keys. Surprisingly, we found that users often negate this security by moving their funds from a multisig address to a regular address and then back again after a period of a few hours to days. We think this happens when users are changing the access control policy on their wallet, although it is unclear why they transfer their funds to a regular address in the interim, and not directly to the new multisig address. This pattern of behavior has led over $12 million dollars to be left insecure over the course of over 22,000 transactions. What users may not appreciate is that the temporary weakening of security is advertised to potential attackers on the blockchain."

**Comment by daira** [comment]
We're considering using RedDSA-based multisignatures instead of changing the circuit. This would have the advantage of being usable with Sapling without further deployment complications.

**Comment by arielgabizon** [comment]
In fact, I've just watch Peiter Wuille's awesome talk  https://www.youtube.com/watch?v=YSUVRj8iznU&t=2031s
It seems given that we already have a Schnorr signature in the pipeline,
we could use similar techniques to do all sorts of smart contracts for shielded payments

**Comment by nathan-at-least** [comment]
Bump, as I just linked to this ticket as background for a Blossom Goal. I've heard that with Sapling's Schnorr signature scheme it's feasible to do threshold signatures with no protocol changes, which might be distinct from this ticket.

The first thing for us to do is gather use case requirements.

**Comment by nathan-at-least** [comment]
cc @arielgabizon who I believe has been looking into threshold signatures and multisig schemes. Feel free to ping others, too.

**Comment by arielgabizon** [comment]
@nathan-at-least I'm working for starters on implementing 2-2 shielded multisig @bitcartel helped me a bit during funweek.

**Comment by mms710** [comment]
Including this in the blossom goals as a possible duplicate of https://github.com/zcash/zcash/issues/3679. Will determine if this is a duplicate after gathering more requirements from partners.
